# AI ANALYST LAB — Python Crash Course

![](../_img/ai_analyst_lab.png)

### From Zero to Python — a gentle companion for people who have never programmed
## Session 5: pandas — your first real data table

Feedback should be sent to [goran.milovanovic@datakolektiv.com](mailto:goran.milovanovic@datakolektiv.com).

Until now, your "data" has been a handful of values you typed by hand. Today that changes. You will download a **real dataset** — **891 real passengers** of the Titanic — from a file on the internet, and load it into a **pandas DataFrame**: the data **table** that sits at the very heart of almost everything in the AI Analyst LAB course.

And you are in for a small surprise. The five passengers you have been carrying since Session 2 were never made up — you will meet the whole ship they came from.

Go slowly. You **cannot break anything**.

***
### What we will do today

- **Make a folder** for the course and **download** the real dataset — step by step, slowly.
- Meet **pandas** and load a data file with a **single line**.
- Meet the **DataFrame** — a real table whose columns can each hold a different kind of value.
- Take the analyst's **"first look"**: `.head()`, `.shape`, `.dtypes`, `.describe()`.
- Pull out **one column** by its name.
- Keep only the **rows that meet a condition** (for example, only the survivors).
- See what **real, messy data** looks like — missing values.

Work in the **IDLE editor**: open a new file (**File → New File**), and — as you'll see in the very first section — save it as **`passengers.py`** in a folder we are about to make. Run it with **Run → Run Module** each time. *(On a Mac, use the menu — the F5 key opens Dictation.)*

***
## 1. Why we need a *table*

Think of a **spreadsheet**, or a ship's old paper **passenger list**: one **row** per person, and a set of **columns** — name, age, ticket class, whether they survived. The quiet magic is that **each column holds a different *kind* of value**: the *name* column is text, the *age* column is numbers, the *survived* column is yes/no.

That is exactly the wall we hit last session. Remember our little story about *kinds of data*:

- a **list** (Session 2) happily **mixes** kinds — text and numbers together;
- a **NumPy array** (Session 4) is **strict** — every value must be the **same** kind.

A real dataset needs **both** at once: a *name* column of text **and** an *age* column of numbers, sitting side by side. Neither a plain list nor a single array can do that cleanly. The answer is a new kind of thing — a **table where each column keeps its own kind** — and that is precisely what **pandas** gives you. Its table is called a **DataFrame**.

> **pandas** (the name comes from *"Python Data Analysis"*) is another **add-on toolbox**, like NumPy — you installed it back in Session 0. By long tradition it is nicknamed **`pd`**.

***
## 2. First, get the data onto your computer

Before we can load data, we need a **data file** — a real file, sitting in a place you can find. We will do this **once**, slowly, in **four small steps**. Follow them in order; don't rush.

*(One quick word: our file is a **CSV** — short for **"comma-separated values"**. It is the plainest possible spreadsheet: a text file where each line is one row, and commas separate the columns. Excel, Google Sheets, and pandas can all open it.)*

### Step 1 — Make a folder for this course

Make a **new folder** on your **Desktop**, and name it **`AI_AnalystLAB`**.

- **On Windows:** right-click an empty spot on the Desktop → **New → Folder**, then type the name.
- **On a Mac:** click the Desktop, then in the top menu **File → New Folder**, and type the name.

Every file you create in this course — your Python files **and** your data files — will live **inside this one folder, together**. That "together" is the whole secret to loading data without headaches, as you'll see in a moment.

### Step 2 — Start today's Python file, and save it in that folder

Open **IDLE**, then **File → New File** to open a fresh editor window. Before you type anything, **save it**: **File → Save**, navigate **into your `AI_AnalystLAB` folder**, and save it with the name **`passengers.py`**.

> **⚠️ Save it *first*, before running anything.** Saving is what tells Python *where "here" is* — and in a moment we will rely on "here" meaning your `AI_AnalystLAB` folder. An unsaved, untitled window has no "here" yet.

### Step 3 — Download the dataset into the *same* folder

**Right-click** the link below and choose **"Save Link As…"** (some browsers word it **"Download Linked File As…"**):

> https://raw.githubusercontent.com/datakolektiv/AI_AnalystLAB/main/_data/titanic/Titanic_train.csv

In the save window that appears, browse **into your `AI_AnalystLAB` folder** and save the file there. Keep its name **exactly** `Titanic_train.csv`.

> **⚠️ A few things that commonly trip people up here — read these first:**
> - **If a plain left-click just opens a page full of text and numbers** — don't worry, that is simply the file's *contents* being shown in the browser. Go back and **right-click** the link instead, then **"Save Link As…"**.
> - **Keep the name exactly `Titanic_train.csv`.** Some browsers try to add `.txt` on the end (making `Titanic_train.csv.txt`). If yours does, **rename the file** to remove the extra `.txt`.
> - **You only do this download once.** From now on the file simply sits in your folder, ready every time.
> - **If any of this snags, do not fight it — contact your course instructor.** This step is a one-time setup, not a test of your patience.

### Step 4 — Check that the file is really there

Open your **`AI_AnalystLAB`** folder and look inside. You should see **two files sitting side by side**:

- `passengers.py`   — your Python file
- `Titanic_train.csv`   — the data

If both are there, you are ready to load real data. If the CSV isn't there, repeat **Step 3** — and remember, your instructor is one message away.

***
## 3. Load it — one line does it

Now the moment it was all for. In your `passengers.py` editor, type these lines. First we bring in the pandas toolbox (nicknamed `pd`); then we read the file **into a table** and peek at the top of it:

In [1]:
import pandas as pd

passengers = pd.read_csv("Titanic_train.csv")
print(passengers.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


That's it — a whole spreadsheet of 891 people, loaded in **one line**. Two things to notice:

- Because `Titanic_train.csv` sits in the **same folder** as `passengers.py`, Python finds it by **name alone** — no long, complicated path needed. *(This is exactly why Step 1 put everything in one folder.)*
- **`.head()`** shows just the **first few rows**, so you can glance at the table and confirm it loaded. It is the first thing every analyst does with new data.

> **⚠️ The one error to expect here — and it's an easy fix.** If you see **`FileNotFoundError`** mentioning `'Titanic_train.csv'`, it almost always means one thing: the CSV is **not in the same folder** as your `passengers.py`. Put the two files together (Step 3), save, and run again. Still stuck? Contact your instructor.
>
> **⚠️ If instead `import pandas` turns red** with **`ModuleNotFoundError`**, pandas was never installed — redo the Session 0 install, or ask your instructor.

***
### A small surprise

Look closely at the very **first row** — passenger number `0`:

**`Braund, Mr. Owen Harris`**, age **22**.

That is **Owen** — the very same passenger whose dictionary you built by hand back in **Session 2** (`owen_dict`). And look at the ages running down those first rows: **22, 38, 26, 35, 35**. Those are **exactly** the five ages you have been typing since Session 2, and turning into a NumPy array in Session 4.

They were never invented. They were the **real first five passengers** of this dataset all along — and now you are holding **all 891** of them. Everything you practised on five hand-typed people, you can now do for the whole ship.

***
## 4. The analyst's "first look" — four quick checks

Whenever an analyst opens a dataset for the first time, they run the **same four checks** before anything else. Make them a habit; you will see them at the start of nearly every session in the main course.

You already did check **1** — **`.head()`**, a peek at the first rows. Here are the other three. First, **how big is the table**, and **what kind is each column**:

In [2]:
print("Size (rows, columns):", passengers.shape)   # a fact — no ()

print()
print("The kind of each column:")
print(passengers.dtypes)                            # a fact — no ()

Size (rows, columns): (891, 12)

The kind of each column:
PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object


- **`.shape`** reports **`(891, 12)`** — **891 rows** (passengers) and **12 columns** (facts about each one).
- **`.dtypes`** reports the **kind** held by each column — and here our whole *kinds of data* story finally pays off. Look: `Name` is **text**, `Age` is a **number** (`float64`), `Survived` is a **number** (`int64`) — **different kinds, one per column**, all in a single table. That is the thing neither a list nor an array could give us.

Notice there are **no parentheses** on `.shape` or `.dtypes`. That's the `( )`-test from Sessions 2 and 4: they are **facts the table simply has** (attributes), not **actions it performs** (methods). Compare `passengers.shape` (a fact, no `()`) with `passengers.head()` (an action — go fetch the first rows, so it takes `()`).

The fourth check, **`.describe()`**, gives a quick numeric summary of **every number column at once** — the descriptive statistics you met in Session 4 (count, mean, min, max, …), now for a whole table in one call:

In [3]:
print(passengers.describe())

       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200  


Each **column** of that summary is one of the number columns; each **row** is a statistic — the count, the average (`mean`), the smallest (`min`), the largest (`max`), and a few in between. In one glance you learn that the average fare was about 32, the oldest passenger was 80, and — reading the `mean` of the `Survived` column — that roughly **38%** of these passengers survived.

> **One honest note:** `.describe()` runs the numbers for **every** column that *looks* like a number, even ones where an average is meaningless — `PassengerId` is just a label, so its "average" tells you nothing. The tool can't know that; **you** do. Reading a summary always means knowing which columns actually deserve one.

***
## 5. Pulling out a single column

Most of the time you don't want the whole table — you want **one column**. You ask for it by its **name**, in square brackets — and if that feels familiar, it should: it is exactly like **looking up a key in a dictionary** from Session 2.

In [4]:
ages = passengers["Age"]     # give me just the Age column
print(ages.head())

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64


`passengers["Age"]` means *"give me the Age column"* — the very same move as `owen_dict["age"]` gave you one passenger's age in Session 2, only now you get the **whole column** for all 891 people.

And a single column behaves like a **labelled NumPy array**: every summary you learned in Session 4 works on it directly. For instance, the average age across the whole ship:

In [5]:
print("Average age on board:", round(passengers["Age"].mean(), 1))

Average age on board: 29.7


One line gives the average age of 891 people. *(You'll see roughly **29.7**. Hold that thought — it quietly skipped some passengers, and the next section explains why.)*

***
## 6. Keeping only the rows you want

Here is the move an analyst reaches for constantly: **keep only the rows that meet a condition**. It is the **boolean mask** from Session 4 — where you wrote `ages[ages >= 30]` to keep certain *values* — now applied to whole **rows of a table**.

You ask a yes/no question of a column, which gives a `True`/`False` for every row; put that inside square brackets, and pandas keeps only the rows where the answer is `True`. Let's keep only the passengers who **survived** (their `Survived` value is `1`):

In [6]:
survivors = passengers[passengers["Survived"] == 1]
print("Number of survivors:", survivors.shape)

Number of survivors: (342, 12)


`survivors.shape` reports **`(342, 12)`** — of the 891 passengers, **342 survived**. Same stencil idea as Session 4, just keeping **rows** instead of single values.

Now combine the two tricks — filter the rows, then take a column's average — to ask a real question: *what was the average age of the survivors?*

In [7]:
print("Survivors' average age:", round(survivors["Age"].mean(), 1))

Survivors' average age: 28.3


Filtering rows and summarising a column, chained together, is the everyday bread-and-butter of data analysis — and you just did it in two short lines.

***
## 7. Real data is messy — missing values

Hand-typed data is always neat. **Real** data has **holes**. Glance back at the first rows from Section 3: the **`Cabin`** column showed **`NaN`** in places. `NaN` is pandas' way of writing **"there is no value here — it's missing."** Real datasets are full of them, and pretending otherwise is how analyses go wrong.

You can **count** the missing values in each column at a glance:

In [8]:
print(passengers.isna().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


The count tells an honest story about this dataset: **177** passengers have **no recorded age**, **687** have no cabin listed, **2** have no port of departure — and the other columns are complete. That is completely normal for real-world data.

This also explains the little mystery from Section 5. When you asked for `passengers["Age"].mean()`, pandas quietly **skipped the 177 missing ages** and averaged the rest — which is why you still got a sensible **29.7**.

> **What we do — and don't do — today.** Right now we only **notice** and **count** missing values. Simply *recognising that they exist* is the first real skill of a data analyst. Deciding what to *do* about them — fill them in, or leave them out — is a proper topic of the **main AI Analyst LAB course**. For now, whenever you see **`NaN`**, read it as **"missing."**

***
## What you can now do — and what's next

Today the passengers stopped being five numbers you typed and became a **real dataset**. You learned to:

- **make a course folder, download a CSV, and load it** with `pd.read_csv("Titanic_train.csv")` — a whole spreadsheet in one line;
- meet the **DataFrame** — a real table where **each column keeps its own kind** of value;
- take the analyst's **first look**: `.head()`, `.shape`, `.dtypes`, `.describe()`;
- **select a column** by name (`passengers["Age"]`) — dictionary lookup, on a table;
- **keep the rows** that meet a condition (`passengers[passengers["Survived"] == 1]`) — the boolean mask, on rows;
- and **recognise missing values** (`NaN`) in real, messy data.

Our story of *kinds of data* is now complete, all three chapters: **lists mix freely (Session 2) → arrays are strict, one kind (Session 4) → a DataFrame keeps one kind per column (today).**

Next, **Session 6 (pandas II)** teaches the analyst's most-used verb: **group and summarise**. In a single line you will answer questions like *"what was the survival rate for women versus men?"* — precisely the kind of result the whole AI Analyst LAB course is built to produce.

> **⚠️ Quick reminders:**
> - Keep your **`.py` file and your data file in the *same folder*.**
> - **`FileNotFoundError`** → the CSV isn't sitting next to your script. Fix that first.
> - **Attributes** (`.shape`, `.dtypes`) take **no** `( )`; **methods** (`.head()`, `.describe()`) do.
> - Read a column by **name in brackets**: `passengers["Age"]` — just like a dictionary key.

See you in Session 6.

<hr>

![](../_img/DK_Logo_White_150.png)

DataKolektiv, 2026.

[hello@datakolektiv.com](mailto:hello@datakolektiv.com)

<font size=1>License: [GPLv3](../LICENSE). This Notebook is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version. This Notebook is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.</font>